template/interpolate/azure/keyvault/keyvault.go

```diff
@@ -0,0 +1,216 @@
+	if oauthConfig == nil {
+		return nil, nil, fmt.Errorf("Unable to configure OAuthConfig for tenant %s", config.TenantID)
+	}
+
+	sender := sender.BuildSender("AzureRM")
+	keyVaultAuth, err := config.GetAuthorizationToken(sender, oauthConfig, env.TokenAudience)
+	if err != nil {
+		return nil, nil, err
+	}
+	endpoint := env.ResourceManagerEndpoint
+
+	VaultsClient := keyvault.NewVaultsClientWithBaseURI(endpoint, config.SubscriptionID)
+	VaultsClient.Authorizer = keyVaultAuth
+
+	ManagementClient := keyvaultmgmt.New()
+	MgmtAuthorizer := config.BearerAuthorizerCallback(sender, oauthConfig)
+	ManagementClient.Client.Authorizer = MgmtAuthorizer
+
+	return &VaultsClient, &ManagementClient, nil
+}
+
+func GetKeyVaultBaseUrlFromID(ctx context.Context, client *keyvault.VaultsClient, keyVaultId string) (string, error) {
+	if keyVaultId == "" {
+		return "", fmt.Errorf("keyVaultId is empty")
+	}
+
+	id, err := ParseAzureResourceID(keyVaultId)
+	if err != nil {
+		return "", err
+	}
+	resourceGroup := id.ResourceGroup
+
+	vaultName, ok := id.Path["vaults"]
+	if !ok {
+		return "", fmt.Errorf("resource id does not contain `vaults`: %q", keyVaultId)
+	}
+
+	resp, err := client.Get(ctx, resourceGroup, vaultName)
+	if err != nil {
+		if resp.Response.StatusCode == http.StatusNotFound {
+			return "", fmt.Errorf("Error unable to find KeyVault %q (Resource Group %q): %+v", vaultName, resourceGroup, err)
+		}
+		return "", fmt.Errorf("Error making Read request on KeyVault %q (Resource Group %q): %+v", vaultName, resourceGroup, err)
+	}
+
+	if resp.Properties == nil || resp.Properties.VaultURI == nil {
+		return "", fmt.Errorf("vault (%s) response properties or VaultURI is nil", keyVaultId)
+	}
+
+	return *resp.Properties.VaultURI, nil
+}
+
+// ParseAzureResourceID converts a long-form Azure Resource Manager ID
+// into a ResourceID. We make assumptions about the structure of URLs,
+// which is obviously not good, but the best thing available given the
+// SDK.
+func ParseAzureResourceID(id string) (*ResourceID, error) {
+	idURL, err := url.ParseRequestURI(id)
+	if err != nil {
+		return nil, fmt.Errorf("Cannot parse Azure ID: %s", err)
+	}
+
+	path := idURL.Path
+
+	path = strings.TrimPrefix(path, "/")
+	path = strings.TrimSuffix(path, "/")
+
+	components := strings.Split(path, "/")
+
+	// We should have an even number of key-value pairs.
+	if len(components)%2 != 0 {
+		return nil, fmt.Errorf("The number of path segments is not divisible by 2 in %q", path)
+	}
+
+	var subscriptionID string
+
+	// Put the constituent key-value pairs into a map
+	componentMap := make(map[string]string, len(components)/2)
+	for current := 0; current < len(components); current += 2 {
+		key := components[current]
+		value := components[current+1]
+
+		// Check key/value for empty strings.
+		if key == "" || value == "" {
+			return nil, fmt.Errorf("Key/Value cannot be empty strings. Key: '%s', Value: '%s'", key, value)
+		}
```

Comment:
It is not clear to me why the initial capacity is divided by two, or why there counter is incremented by two. Maybe adding a comment explaining what is going here could help.

Reply:
To be honest, I copied that function from the azurerm provider but I totally agree the code is not clear! 

In fact, I actually struggled a bit following it so I am not surprised you did as well, specially if you are not familiar with Azure and the way they construct their resource ids

I will add some comments on the code and see if it becomes a bit more clear